state/protocol/protocol_state/kvstore/encoding.go

```diff
@@ -0,0 +1,49 @@
+package kvstore
+
+import (
+	"bytes"
+
+	"github.com/onflow/flow-go/module/irrecoverable"
+
+	"github.com/vmihailenco/msgpack/v4"
+)
+
+// VersionedEncodable defines the interface for a versioned key-value store independent
+// of the set of keys which are supported. This allows the storage layer to support
+// storing different key-value model versions within the same software version.
+type VersionedEncodable interface {
```

Comment:
Have you considered using similar approach to what we have with fingerprint? 
Instead of defining a must have function for each model we could use default serialization using `msgpack` and use an optional method to override the behavior when needed. 
Not saying we should do it but it seems beneficial to do it instead of implementing the same method in each model version(probably even with the same logic in each method).

Reply:
I do like this idea, spent a bit of time on it but had a hard time getting it to work well in this case. 

`Fingerprint` accepts any input, though we assume it is a struct. Here we would accept an instance of the public interfaces. Marshalling does work in this case (it marshals based on the underlying type implementing the interface) but for me it makes the resulting code less clear. Overall I don't think it's worth it for this case.